javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
+            AllWrittenFlags & ~LeftAlign)
+        rejectPrecision()
+        if (flags.leftAlign && width < 0)
+          throw new MissingFormatWidthException("%-%")
+        padAndSendToDest(flags, width, "%")
+
+      case 'n' =>
+        validateFlagsForPercentAndNewline(flags, conversion, AllWrittenFlags)
+        rejectPrecision()
+        if (width >= 0)
+          throw new IllegalFormatWidthException(width)
+        sendToDest("\n")
 
-          def sciNotation(precision: Int) = {
-            val exp = numberArg.toExponential(precision)
-            with_+({
-              // check if we need additional 0 padding in exponent
-              // JavaDoc: at least 2 digits
-              if ('e' == exp.charAt(exp.length - 3)) {
-                exp.substring(0, exp.length - 1) + "0" +
-                  exp.charAt(exp.length - 1)
-              } else exp
-            }, numberArg.isNaN || numberArg.isInfinite)
-          }
+      case _ =>
+        throw new UnknownFormatConversionException(conversion.toString)
+    }
+  }
+
+  @inline private def validateFlags(flags: Flags, conversion: Char,
+      invalidFlags: Int): Unit = {
+
+    @noinline def flagsConversionMismatch(): Nothing = {
+      throw new FormatFlagsConversionMismatchException(
+          flagsToString(new Flags(flags.bits & invalidFlags)), conversion)
+    }
+
+    if ((flags.bits & invalidFlags) != 0)
+      flagsConversionMismatch()
+
+    @noinline def illegalFlags(): Nothing =
+      throw new IllegalFormatFlagsException(flagsToString(flags))
+
+    /* The test `(invalidFlags & BadCombo) == 0` is redundant, but is
+     * constant-folded away at called site, and if false it allows to dce the
+     * test after the `&&`. If both tests are eliminated, the entire `if`
+     * disappears.
+     */
+    val BadCombo1 = LeftAlign | ZeroPad
+    val BadCombo2 = PositivePlus | PositiveSpace
+    if (((invalidFlags & BadCombo1) == 0 && (flags.bits & BadCombo1) == BadCombo1) ||
+        ((invalidFlags & BadCombo2) == 0 && (flags.bits & BadCombo2) == BadCombo2)) {
+      illegalFlags()
+    }
+  }
+
+  @inline private def validateFlagsForPercentAndNewline(flags: Flags,
+      conversion: Char, invalidFlags: Int): Unit = {
+
+    @noinline def illegalFlags(): Nothing =
+      throw new IllegalFormatFlagsException(flagsToString(flags))
+
+    if ((flags.bits & invalidFlags) != 0)
+      illegalFlags()
+  }
+
+  private def flagsToString(flags: Flags): String = {
+    (if (flags.leftAlign) "-" else "") +
+    (if (flags.altFormat) "#" else "") +
+    (if (flags.positivePlus) "+" else "") +
+    (if (flags.positiveSpace) " " else "") +
+    (if (flags.zeroPad) "0" else "") +
+    (if (flags.useGroupingSeps) "," else "") +
+    (if (flags.negativeParen) "(" else "") +
+    (if (flags.useLastIndex) "<" else "")
+  }
+
+  private def computerizedScientificNotation(x: Double, precision: Int,
+      forceDecimalSep: Boolean): String = {
```

Comment:
call this `altFormat`? Otherwise the logic that `altFormat` means `forceDecimalSep` for efg is very implicit.

Reply:
I found it was more explicit to name this `forceDecimalSep` here. It is semantically related to the "notation" thing. `altFormat` is a very generic thing which applies to many different conversions, and takes on different *meanings* depending on the conversion. Once we're inside `xyzNotation`, there is a more specific meaning, with a more specific name.

I have made the `forceDecimalSep = altFormat` more explicit at call site (in `efgCommon`) to compensate.